braindecode/models/eegminer.py

```diff
@@ -0,0 +1,230 @@
+            clamp_f_mean=self.clamp_f_mean,
+        )
+
+        # Forward method
+        if self.method == "mag":
+            self.method_forward = self._apply_mag_forward
+            self.n_features = self.n_chans * self.n_filters
+            self.ensure_dim = nn.Identity()
+        elif self.method == "corr":
+            self.method_forward = partial(
+                self._apply_corr_forward,
+                n_chans=self.n_chans,
+                n_filters=self.n_filters,
+                n_times=self.n_times,
+            )
+            self.n_features = self.n_filters * self.n_chans * (self.n_chans - 1) // 2
+            self.ensure_dim = nn.Identity()
+        elif self.method == "plv":
+            self.method_forward = partial(self._apply_plv, n_chans=self.n_chans)
+            self.ensure_dim = Rearrange("... d -> ... 1 d")
+            self.n_features = self.n_filters * self.n_chans * (self.n_chans - 1) // 2
+
+        # Classifier
+        self.batch_layer = nn.BatchNorm1d(self.n_features, affine=False)
+        self.final_layer = nn.Linear(self.n_features, self.n_outputs)
+        nn.init.zeros_(self.final_layer.bias)
+
+    def forward(self, x):
+        """x: (batch, electrodes, time)"""
+        batch = x.shape[0]
+        x = self.ensure_dim(x)
+        # Apply Gaussian filters in frequency domain
+        # x -> (batch, electrodes * filters, time)
+        x = self.filter(x)
+
+        x = self.method_forward(x=x, batch=batch)
+        # Classifier
+        # Note that the order of dimensions before flattening the feature vector is important
+        # for attributing feature weights during interpretation.
+        x = x.reshape(batch, self.n_features)
+        x = self.batch_layer(x)
+        x = self.final_layer(x)
+
+        return x
+
+    @staticmethod
+    def _apply_mag_forward(x, batch=None):
+        # Signal magnitude
+        x = x * x
+        x = x.mean(dim=-1)
+        x = torch.sqrt(x)
+        return x
+
+    @staticmethod
+    def _apply_corr_forward(x, batch, n_chans, n_filters, n_times):
+        x = x.reshape(batch, n_chans, n_filters, n_times).transpose(-3, -2)
+        x = (x - x.mean(dim=-1, keepdim=True)) / torch.sqrt(
+            x.var(dim=-1, keepdim=True) + 1e-6
+        )
+        x = torch.matmul(x, x.transpose(-2, -1)) / x.shape[-1]
+        x = x.transpose(-3, -2).transpose(-2, -1)  # move filter channels to the end
+        x = x.abs()
+
+        # Get upper triu of symmetric connectivity matrix
+        triu = torch.triu_indices(n_chans, n_chans, 1)
+        x = x[:, triu[0], triu[1], :]
+
+        return x
+
+    @staticmethod
+    def _apply_plv(x, n_chans, batch=None):
+        # Compute PLV connectivity
+        # x -> (batch, electrodes, electrodes, filters)
+        x = x.transpose(-4, -3)  # swap electrodes and filters
+        x = F.plv_time(x, forward_fourier=False)
+        x = x.transpose(-3, -2).transpose(-2, -1)  # move filter channels to the end
```

Comment:
Same here, is it possible to avoid the 2 transposes?

Reply:
done @sylvchev